protocol/contracts/GenericZap.sol

```diff
@@ -0,0 +1,918 @@
+  uint256 private constant DEADLINE = 0xf000000000000000000000000000000000000000000000000000000000000000;
+
+  struct ZapLiquidityRequest {
+    uint256 firstSwapMinAmountOut;
+    uint248 poolSwapMinAmountOut;
+    bool isOneSidedLiquidityAddition;
+    address otherToken;
+    bytes poolSwapData;
+  }
+
+  event ZappedIn(address indexed sender, address fromToken, uint256 fromAmount, address toToken, uint256 amountOut);
+  event UniswapV2RouterSet(address router);
+  event ZappedLPUniV2(address indexed recipient, address token0, address token1, uint256 amountA, uint256 amountB);
+  event TokenRecovered(address token, address to, uint256 amount);
+  event ZappedLPCurve(address indexed recipient, address fromToken, uint256 liquidity, uint256[] amounts);
+  event ZappedLiquidityBalancerPool(address indexed recipient, address fromToken, uint256 fromAmount, uint256[] maxAmountsIn);
+
+  constructor (
+    address _router
+  ) {
+    uniswapV2Router = IUniswapV2Router(_router);
+  }
+
+  /**
+   * @notice recover token or ETH
+   * @param _token token to recover
+   * @param _to receiver of recovered token
+   * @param _amount amount to recover
+   */
+  function recoverToken(address _token, address _to, uint256 _amount) external onlyOwner {
+    require(_to != address(0), "Invalid receiver");
+    if (_token == address(0)) {
+      // this is effectively how OpenZeppelin transfers eth
+      require(address(this).balance >= _amount, "Address: insufficient balance");
+      (bool success,) = _to.call{value: _amount}(""); 
+      require(success, "Address: unable to send value");
+    } else {
+      _transferToken(IERC20(_token), _to, _amount);
+    }
+    
+    emit TokenRecovered(_token, _to, _amount);
+  }
+
+  function setUniswapV2Router(address _router) external onlyOwner {
+    uniswapV2Router = IUniswapV2Router(_router);
+    emit UniswapV2RouterSet(_router);
+  }
+
+  /**
+   * @notice This function zaps ETH and ERC20 tokens to Temple token
+   * @param fromToken The token used for entry (address(0) if ether)
+   * @param fromAmount The amount of fromToken to zap
+   * @param toToken Exit token
+   * @param amountOutMin The minimum acceptable quantity of TEMPLE to receive
+   * @param swapTarget Execution target for the swap
+   * @param swapData DEX data
+   * @return amountOut Amount of tokens received
+   */
+  function zapIn(
+    address fromToken,
+    uint256 fromAmount,
+    address toToken,
+    uint256 amountOutMin,
+    address swapTarget,
+    bytes calldata swapData
+  ) external payable whenNotPaused returns (uint256 amountOut) {
+    require(approvedTargets[fromToken][swapTarget] == true, "Unsupported token/target");
+
+    amountOut = _zapIn(
+      fromToken,
+      fromAmount,
+      toToken,
+      amountOutMin,
+      swapTarget,
+      swapData
+    );
+  }
+
+  function zapLiquidityBalancerPoolFor(
+    address _fromToken,
```

Comment:
Wonder if it's worth breaking up the balancer | curve | uni into separate implementation files? Is this starting to hit up against the bytesize limits or ok?

Reply:
You can run `yarn size` and it'll print out the size of all the contracts - was super useful when splitting up OpsManager in the core contracts.

Reply:
`yarn size` shows `24.052` . very close. will do some splitting